### PR TITLE
apps/cms: Clean up order of options in help output and documentation

### PR DESCRIPTION
**Pull request by DDvO**
This brings the CLI options of the CMS app in a more sensible order.

To this end made the `opt_` API more flexible:
* APPS: Allow duplicate entries in options list, marking them `OPT_DUP`
* APPS: Allow non-option parameters appear anywhere in options list, marking them `OPT_PARAM`

such that
* the `-signer` option can be occur both in the Signing and Verification sections, likewise the `-cades` option,
while adequately differentiating their semantics in each of these sections
* the non-option `cert...` parameters can be described in the section they belong to: Encryption and decryption.

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

**Comment by richsalz** [comment]
Wow, thanks for tackling this!  `OPT_DUP` is clever.

There are going to be conflicts between this and #14182, which moves the cipher-fetching/lookup *after* the processing of all other options (-provider etc).  I know that #14182 is required for beta.  Would you consider rebasing this, not, on top of that PR?

**Comment by DDvO** [comment]
> Wow, thanks for tackling this! `OPT_DUP` is clever.

Pleased to hear!

> There are going to be conflicts between this and #14182, which moves the cipher-fetching/lookup _after_ the processing of all other options (-provider etc). I know that #14182 is required for beta. Would you consider rebasing this, not, on top of that PR?

Could do that, but there is a simpler solution:
I've just added an approval by me on that PR.
So after that is merged a day later, I can directly rebase this PR on your changes.
The merge conflicts will not be hard to solve.

**Comment by richsalz** [comment]
> The merge conflicts will not be hard to solve.

Reach out if anything's confusing/problematic.

File: apps/cms.c
Comment (FdaSilvaYY) [inline apps/cms.c:51]: Look like a Dormant  bug:  16 in decimal is equal to 0x10, no ?
So it is equal to SMIME_OP ?

File: apps/cms.c
Comment (FdaSilvaYY) [inline apps/cms.c:174]: a bunch of  these descriptions were previously indented, when they are on a new line.

**Review by FdaSilvaYY — CHANGES_REQUESTED** [review]
great work 👍

File: apps/cms.c
Comment (DDvO) [reply apps/cms.c:175]: Removed those indentations for consistency (and to silence `check-format.pl` on them).

File: apps/cms.c
Comment (DDvO) [reply apps/cms.c]: Yep, done.

File: apps/cms.c
Comment (DDvO) [reply apps/cms.c]: Done

File: apps/cms.c
Comment (DDvO) [reply apps/cms.c:51]: Good spotting!
In this case the clash did not do harm because ` (16 | SMIME_IP | SMIME_OP)` has the same value as ` (0 | SMIME_IP | SMIME_OP)` which is not taken.
Yet in case a new value (such as `17`) is added there will be an issue.
So changed the definition of the flags to
```
#define SMIME_OP                0x100
#define SMIME_IP                0x200
#define SMIME_SIGNERS           0x400
```

**Comment by DDvO** [comment]
> great work +1

Thanks @FdaSilvaYY, also for your comments :+1:

**Comment by DDvO** [comment]
Fixed again merge conflicts.
Can we now finalize this?

**Comment by t8m** [comment]
Needs a rebase again.

**Comment by DDvO** [comment]
> Needs a rebase again.

Thx for the notice - done.

File: doc/man1/openssl-cms.pod.in
Comment (t8m) [inline doc/man1/openssl-cms.pod.in:847]: `an encrypted message`

**Review by t8m — CHANGES_REQUESTED** [review]
Some nits

File: apps/cms.c
Comment (DDvO) [reply apps/cms.c]: I suspected that the order of the option `enum` elements partially matters, but may be wrong on this.
And for maintaining the options I think it may help to keep them consistent, but it also adds a burden.
So reverted the changed order as you suggested.

File: apps/cms.c
Comment (DDvO) [reply apps/cms.c]: Oops, but this gone now anyway, due to the reverted change.

File: apps/cms.c
Comment (DDvO) [reply apps/cms.c]: Done

File: apps/cms.c
Comment (DDvO) [reply apps/cms.c]: In the wider sense yes, since `-keyopt` influences the signing and encryption output.
Yet it may be more intuitive to group it with the "keys and passwords" options - moved it there.

File: doc/man1/openssl-cms.pod.in
Comment (DDvO) [reply doc/man1/openssl-cms.pod.in]: Done

File: doc/man1/openssl-cms.pod.in
Comment (DDvO) [reply doc/man1/openssl-cms.pod.in]: Done

File: doc/man1/openssl-cms.pod.in
Comment (DDvO) [reply doc/man1/openssl-cms.pod.in:847]: Done

**Review by t8m — APPROVED** [review]
Approved if CI passes.

**Comment by DDvO** [comment]
Merged - thanks @FdaSilvaYY and @t8m

**Pull request closed without merging**
